### PR TITLE
[autograd Function] Don't materialize forward grad for non-differentiable types

### PR DESCRIPTION
**Pull request by soulitzer**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #91222
* #91214
* __->__ #91183

### CODE REVIEW

File: torch/csrc/autograd/python_function.cpp [hunk 1]
```diff
@@ -392,7 +392,8 @@ static void _wrap_outputs(
     for (const auto i : c10::irange(num_inputs)) {
       PyObject* input = nullptr;
       if (self->is_variable_input[i]) {
-        if (grad_inputs[variable_idx].defined() || !self->materialize_grads) {
+        if (grad_inputs[variable_idx].defined() || !self->materialize_grads ||
```
Comment (zou3519) [inline torch/csrc/autograd/python_function.cpp:395]: Need to update the docs of ctx.set_materialize_grads. It doesn't say anything about the JVP case: https://pytorch.org/docs/stable/generated/torch.autograd.function.FunctionCtx.set_materialize_grads.html

**Review by zou3519 — APPROVED** [review]
LGTM. Needs a rebase though, https://github.com/pytorch/pytorch/blob/f8b348c1fc964ea12b39f37538c4e6fff1d752dc/test/functorch/test_ops.py#L1485 should pass after this change

File: torch/csrc/autograd/python_function.cpp (hunk 1 above)
Comment (soulitzer) [reply torch/csrc/autograd/python_function.cpp:395]: Updated.

**Pull request closed without merging**